nilearn/connectivity/connectivity_matrices.py

```diff
@@ -0,0 +1,341 @@
+        _check_matrix(matrix, 'square')
+        if matrix.shape[0] != n_features:
+            raise ValueError("Matrices are not of the same shape.")
+        _check_matrix(matrix, 'spd')
+
+    # Initialization
+    matrices = np.array(matrices)
+    if init is None:
+        gmean = np.mean(matrices, axis=0)
+    else:
+        _check_matrix(init, 'square')
+        if init.shape[0] != n_features:
+            raise ValueError("Initialization has incorrect shape.")
+        _check_matrix(init, 'spd')
+        gmean = init
+
+    norm_old = np.inf
+    step = 1.
+
+    # Gradient descent
+    for n in range(max_iter):
+        # Computation of the gradient
+        vals_gmean, vecs_gmean = linalg.eigh(gmean)
+        gmean_inv_sqrt = _form_symmetric(np.sqrt, 1. / vals_gmean, vecs_gmean)
+        whitened_matrices = [gmean_inv_sqrt.dot(matrix).dot(gmean_inv_sqrt)
+                             for matrix in matrices]
+        logs = [_map_eigenvalues(np.log, w_mat) for w_mat in whitened_matrices]
+        logs_mean = np.mean(logs, axis=0)  # Covariant derivative is
+                                           # - gmean.dot(logms_mean)
+        if np.any(np.isnan(logs_mean)):
+            raise FloatingPointError("Nan value after logarithm operation.")
+
+        norm = np.linalg.norm(logs_mean)  # Norm of the covariant derivative on
+                                          # the tangent space at point gmean
+
+        # Update of the minimizer
+        vals_log, vecs_log = linalg.eigh(logs_mean)
+        gmean_sqrt = _form_symmetric(np.sqrt, vals_gmean, vecs_gmean)
+        # Move along the geodesic
+        gmean = gmean_sqrt.dot(
+            _form_symmetric(np.exp, vals_log * step, vecs_log)).dot(gmean_sqrt)
+
+        # Update the norm and the step size
+        if norm < norm_old:
+            norm_old = norm
+        if norm > norm_old:
+            step = step / 2.
+            norm = norm_old
+        if tol is not None and norm / gmean.size < tol:
+            break
+    if tol is not None and norm / gmean.size >= tol:
+        warnings.warn("Maximum number of iterations {0} reached without "
+                      "getting to the requested tolerance level "
+                      "{1}.".format(max_iter, tol))
+
+    return gmean
+
+
+def sym_to_vec(symmetric):
+    """Return the flattened lower triangular part of an array.
+
+    Acts on the last two dimensions of the array if not 2-dimensional.
+
+    Parameters
+    ----------
+    symmetric : numpy.ndarray, shape (..., n_features, n_features)
+        Input array.
+
+    Returns
+    -------
+    output : numpy.ndarray, shape (..., n_features * (n_features + 1) / 2)
+        The output flattened lower triangular part of symmetric.
+    """
+    tril_mask = np.tril(np.ones(symmetric.shape[-2:]), -1).astype(np.bool)
+    symmetric = symmetric.copy()
+    symmetric[..., tril_mask] *= sqrt(2)
+    tril_mask.flat[::symmetric.shape[-1] + 1] = True
+    return symmetric[..., tril_mask]
```

Comment:
I may be wrong, but there must be a simpler way to implement this function, in particular without having to copy the input array (`symmetric = symmetric.copy()` l.205 of the current diff).